{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div style=\"color: rgb(27,94,32); background: rgb(200,230,201); border: solid 1px rgb(129,199,132); padding: 10px;\">\n",
    "\n",
    "# Week 10: Optimisation\n",
    "\n",
    "In this lab we'll use simulated annealing to perform network alignment. \n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---------------\n",
    "\n",
    "## Introduction\n",
    "\n",
    "**What is a network?**\n",
    "\n",
    "Biological systems are complex. \n",
    "Proteins (and other functional elements) encoded in DNA do not work in isolation. <br>\n",
    "Rather, their existence is contextual within a larger, highly dynamic biological system. \n",
    "\n",
    "To model such systems in computers, we use graphs (commonly referred to as 'networks').  <br>\n",
    "Nodes within these networks represent biological entities, such as genes or proteins. <br>\n",
    "Edge represent a relationship, such as regulation (GRN) or binding partner (PPI). \n",
    "\n",
    "Common examples include:\n",
    "- Gene Regulatory Networks (GRNs), which model gene regulation, and \n",
    "- Protein-Protein Interaction (PPI) networks, which model how proteins interact with each other.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "**Our Task**\n",
    "\n",
    "The goal of network alignment is similar to sequence alignment. \n",
    "Consider a situation where there is a model (well-studied) organism, and a newly analysed organism. \n",
    "A typical sequence alignment task is to align predicted genes from the new organism to the model organism (acting as reference) so that functional annotations can be carried over. This hinges on the idea that similar sequences have similar function, and if employed, rapidly accelerates our understanding of the new organism. \n",
    "\n",
    "Network alignment can be considered a similar problem. \n",
    "Instead of transferring annotations of single genes, we could compare their biological networks. \n",
    "This would provide insight into how these organisms differ at a systems level, providing deeper insight. \n",
    "\n",
    "<img src=\"https://raw.githubusercontent.com/melbournebioinformatics/COMP90014_2024/master/tutorials/media/week10/network_alignment.png\" width=\"400\"><br>\n",
    "<small>Adapted from [Cheng-Yu Maa and Chung-Shou Liaob, 2020](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7533294/).</small>\n",
    "<br>\n",
    "\n",
    "Network alignments are represented as a mapping between two networks.<br>\n",
    "Specifically, this mapping is a set of node pairs, where each pair maps a node in network 1 to its counterpart in network 2. \n",
    "\n",
    "The figure above illustrates such an alignment.  <br>\n",
    "The blue network is considered the 'reference', and the orange network the 'query'.<br>\n",
    "Red dotted lines represent the node mapping we wish to obtain. <br>\n",
    "\n",
    "Note that some nodes may not appear in both networks. For example, the reference organism has a gene which is absent from the query organism.\n",
    "This can be seen in the figure above, where the query network has a node 'k' which is missing from the reference network. In addition, some nodes may map to multiple counterpart nodes. For example, the query organism had a gene duplication event, resulting in two genes with similar function. \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "**Why is this tricky?**\n",
    "\n",
    "> Note <br>\n",
    "> In this tutorial we will only investigate global alignment of two networks. <br>\n",
    "> This is (arguably) the simplest problem in the field. \n",
    "\n",
    "Unfortunately, our current technology has limitations. \n",
    "The experimental techniques (ChIP-Seq, yeast-two-hybrid, and scRNA-seq) used to probe biological systems are not perfect. \n",
    "Neither are the computation methods used to impute networks from the experimental data. \n",
    "\n",
    "This results in networks being somewhat inaccurate. \n",
    "For example, yeast-two-hybrid experiments can have up to ~20% false positives/negatives [cite]. \n",
    "A PPI network obtained from this data could therefore have 20% of all edges being incorrect. \n",
    "\n",
    "Think back to network alignment, we can now appreciate the difficulty of this task. \n",
    "Two PPI networks, both generated using yeast-two-hybrid data, are likely to differ drastically. \n",
    "The best alignment of these two networks would therefore try to maximise both sequence and topological similarity.\n",
    "- Sequence similarity: counterpart nodes should have similar sequence (and therefore similar molecular function). \n",
    "- Topological similarity: the aligned networks should have similar topology (and therefore operate similarly at a system level). \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Optimisation for Network Alignment**\n",
    "\n",
    "The goal of optimisation is to find the best solution to a task. \n",
    "It can be used when the task is computationally challenging (eg NP-hard problems), or when there is no clear rule-based way to identify the best solution. \n",
    "\n",
    "The main ingredients are as follows:\n",
    "- Generate solution(s)\n",
    "- Score fitness (objective function)\n",
    "- Return solution with best fitness.\n",
    "\n",
    "Network alignment can be solved (in reasonable time) using an optimisation approach called simulated annealing.  <br>\n",
    "The overall process is iterative:\n",
    "- Start with a random solution and score its fitness. \n",
    "- Generate a neighbour (similar) solution randomly and score its fitness. \n",
    "- If the neighbour solution has better fitness, jump to this solution.\n",
    "\n",
    "So that we don't get stuck in a local optimum, we will add an element of randomness. \n",
    "Sometimes we will jump to worse solutions, hoping they will eventually lead to a global optimum solution. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "--------------\n",
    "\n",
    "## Data and Helper Functions\n",
    "\n",
    "**The Data**\n",
    "\n",
    "Two files are needed for this tutorial. <br>\n",
    "These define the networks shown in the figure above. \n",
    "- https://github.com/melbournebioinformatics/COMP90014_2024/blob/master/tutorials/data/week10/ppi1.adjacency?raw=true\n",
    "- https://github.com/melbournebioinformatics/COMP90014_2024/blob/master/tutorials/data/week10/ppi2.adjacency?raw=true\n",
    "\n",
    "Download these files and place them in a <small>`data/`</small> folder relative to your current working directory. \n",
    "\n",
    "Run the cell below to load and visualise these two networks. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx \n",
    "import matplotlib.pyplot as plt \n",
    "\n",
    "def load_network(filepath: str) -> nx.Graph:\n",
    "    G = nx.Graph()\n",
    "    with open(filepath, 'r') as fp:\n",
    "        next(fp)\n",
    "        line = fp.readline().strip()\n",
    "        while line:\n",
    "            node, sequence, neighbours = line.split('\\t')\n",
    "            G.add_node(node, sequence=sequence)\n",
    "            for neighbour in neighbours.split(','):\n",
    "                G.add_edge(node, neighbour)\n",
    "            line = fp.readline().strip()\n",
    "    return G \n",
    "\n",
    "def render_networks(R: nx.Graph, Q: nx.Graph) -> None:\n",
    "    fig, ax = plt.subplots(nrows=1, ncols=2, figsize=(8, 4))\n",
    "    pos1 = nx.spring_layout(R, k=0.5)\n",
    "    pos2 = nx.spring_layout(Q, k=0.5)\n",
    "    nx.draw(R, pos=pos1, ax=ax[0], with_labels=True, node_size=300, node_color='lightskyblue')\n",
    "    nx.draw(Q, pos=pos2, ax=ax[1], with_labels=True, node_size=300, node_color='salmon')\n",
    "    ax[0].set_title('Reference Network')\n",
    "    ax[1].set_title('Query Network')\n",
    "    plt.show()\n",
    "\n",
    "R = load_network('data/ppi1.adjacency') # reference network\n",
    "Q = load_network('data/ppi2.adjacency') # query network\n",
    "render_networks(R, Q)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Objective Function**\n",
    "\n",
    "An objective function is required to assess the 'fitness' of a solution. <br>\n",
    "This is required so that we can migrate towards better solutions over time. \n",
    "\n",
    "Today we will provide the objective function, which we need to ***minimise***.<br>\n",
    "For your interest, our objective function combines two metrics to return a fitness score (referred to as 'energy'):\n",
    "- *Sequence similarity*, measured by edit distance (i.e. sequence identity with gaps)\n",
    "- *Toplogical similarity*, measured by edge correctness (i.e. % of edges in the reference network which appear in query network) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Levenshtein\n",
    "\n",
    "def objective_function(soln: list, R: nx.Graph, Q: nx.Graph) -> float:\n",
    "    \"\"\"\n",
    "    Calculates 'fitness' using sequence and topological similarity. \n",
    "    Returned value is a measure of dissimilarity between 0 and 1.\n",
    "    Eg a perfect solution returns a value of 0. \n",
    "    \"\"\"\n",
    "    # sequence similarity (using edit distance)\n",
    "    obs_edits, pos_edits = 0, 0\n",
    "    for r_node, q_node in soln:\n",
    "        if r_node is None or q_node is None:\n",
    "            continue\n",
    "        else:\n",
    "            seq1, seq2 = R.nodes[r_node]['sequence'], Q.nodes[q_node]['sequence']\n",
    "            obs_edits += Levenshtein.distance(seq1, seq2)\n",
    "            pos_edits += max([len(seq1), len(seq2)])\n",
    "    sequence_distance = obs_edits / pos_edits\n",
    "\n",
    "    # topological similarity (using Edge Correctness, EC)\n",
    "    graph2_mapper = {r_node: q_node for r_node, q_node in soln if r_node is not None}\n",
    "    total_edges = len(list(R.edges()))\n",
    "    mapped_edges = 0\n",
    "    for r_node1, r_node2 in R.edges():\n",
    "        q_node1, q_node2 = graph2_mapper[r_node1], graph2_mapper[r_node2]\n",
    "        if Q.has_edge(q_node1, q_node2):\n",
    "            mapped_edges += 1\n",
    "    topological_distance = 1 - (mapped_edges / total_edges)\n",
    "\n",
    "    return (sequence_distance + topological_distance) / 2\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---------------\n",
    "\n",
    "## Exercises\n",
    "\n",
    "**Neighbour Solutions**\n",
    "\n",
    "The cell below displays the fitness of two different solutions. \n",
    "\n",
    "We can see that solution 1 (the 'correct' solution) has a lower energy than solution 2. <br>\n",
    "This property of the objective function is needed so we can perform optimisation. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# best solution\n",
    "soln1 = [\n",
    "    ['A', 'A`'], ['B', 'B`'], ['C', 'C`'], ['D', 'D`'], ['E', 'E`'],\n",
    "    ['F', 'F`'], ['G', 'G`'], ['H', 'H`'], ['I', 'I`'], ['J', 'J`'], [None, 'K']\n",
    "]\n",
    "energy = objective_function(soln1, R, Q)\n",
    "print('\\nsolution 1 (correct) ---')\n",
    "print(f'energy={energy:0.2f}')\n",
    "\n",
    "# neighbour solution\n",
    "soln2 = [ \n",
    "    ['A', 'B`'], ['B', 'A`'], ['C', 'C`'], ['D', 'D`'], ['E', 'E`'],\n",
    "    ['F', 'F`'], ['G', 'G`'], ['H', 'H`'], ['I', 'I`'], ['J', 'J`'], [None, 'K']\n",
    "]\n",
    "print('\\nsolution 2 (swapped A and B) ---')\n",
    "energy = objective_function(soln2, R, Q)\n",
    "print(f'energy={energy:0.2f}')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These two solutions are quite similar. The only difference is that two node mappings have been swapped. \n",
    "\n",
    "```python\n",
    "current\t\tneighbour\n",
    "['A', 'A`']\t['A', 'B`'] <-\n",
    "['B', 'B`']\t['B', 'A`'] <-\n",
    "['C', 'C`']\t['C', 'C`']\n",
    "['D', 'D`']\t['D', 'D`']\n",
    "['E', 'E`']\t['E', 'E`']\n",
    "['F', 'F`']\t['F', 'F`']\n",
    "['G', 'G`']\t['G', 'G`']\n",
    "['H', 'H`']\t['H', 'H`']\n",
    "['I', 'I`']\t['I', 'I`']\n",
    "['J', 'J`']\t['J', 'J`']\n",
    "[None, 'K']\t[None, 'K']\n",
    "\n",
    "```\n",
    "\n",
    "We will use this approach to generate 'neighbour' solutions. <br>\n",
    "Starting with a current solution, we will pick two node mappings at random, then perform the swap. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div style=\"color: rgb(27,94,32); background: rgb(200,230,201); border: solid 1px rgb(129,199,132); padding: 10px;\">\n",
    "\n",
    "<h3>Exercise 1</h3>\n",
    "    \n",
    "<b>Challenge:</b> Write a function which generates a neighbour solution. \n",
    "\n",
    "- [ ] Input: the current solution (list of lists).\n",
    "- [ ] Output: a random neighbouring solution (list of lists).\n",
    "\n",
    "Instructions:\n",
    "- Randomly choose two node mappings and swap them. \n",
    "- Both the current and neighbour solution should have the same number of elements. \n",
    "\n",
    "\n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "from copy import deepcopy\n",
    "\n",
    "def generate_neighbour(c_solution: list) -> list:\n",
    "    ### BEGIN SOLUTION\n",
    "    # randomly pick which node mappings to swap\n",
    "    idx1 = random.randint(0, len(c_solution)-1)\n",
    "    idx2 = random.randint(0, len(c_solution)-1)\n",
    "    while idx2 == idx1:\n",
    "        idx2 = random.randint(0, len(c_solution)-1)\n",
    "    \n",
    "    # swap the mappings\n",
    "    n_solution = deepcopy(c_solution)\n",
    "    n_solution[idx1][1], n_solution[idx2][1] = n_solution[idx2][1], n_solution[idx1][1]\n",
    "    return n_solution\n",
    "    ### END SOLUTION\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TESTS \n",
    "# run this cell multiple times. \n",
    "# the new solution should be different each time. \n",
    "# check visually whether your code works (altered pairs displayed in red). \n",
    "c_solution = [\n",
    "    ['A', 'A`'], ['B', 'B`'], ['C', 'C`'], ['D', 'D`'], ['E', 'E`'], \n",
    "    ['F', 'F`'], ['G', 'G`'], ['H', 'H`'], ['I', 'I`'], ['J', 'J`'], [None, 'K']\n",
    "]\n",
    "n_solution = generate_neighbour(c_solution)\n",
    "print('current\\t\\tneighbour')\n",
    "for c_pair, n_pair in zip(c_solution, n_solution):\n",
    "    if c_pair == n_pair:\n",
    "        print(f'{c_pair}\\t{n_pair}')\n",
    "    else:\n",
    "        print(f\"\\x1b[31m{c_pair}\\t{n_pair}\\x1b[0m\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Acceptance Probabilities**\n",
    "\n",
    "Once we have generated a new solution, we need to make a decision: whether to jump or stay. <br>\n",
    "This decision will be represented as a probability. \n",
    "\n",
    "There are three things to consider:\n",
    "1. If the new solution is better (lower energy), we should always accept it. \n",
    "2. If the new solution is worse, we should sometimes accept it, so that we don't get stuck in a local optimum. \n",
    "3. As we get closer to the end of our run, we should accept worse solutions less often.  \n",
    "\n",
    "This process (or 'schedule') means we will become greedier over time.<br>\n",
    "We will define a 'temperature' variable so we can take this into consideration. \n",
    "\n",
    "To calculate the acceptance probability we will use the following formula. \n",
    "\n",
    "$P(e, e', T)=1$ if $e'<e$, otherwise $\\exp(-(e'-e)/T)$\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div style=\"color: rgb(27,94,32); background: rgb(200,230,201); border: solid 1px rgb(129,199,132); padding: 10px;\">\n",
    "\n",
    "<h3>Exercise 2</h3>\n",
    "    \n",
    "<b>Challenge:</b> Write a function which calculates the acceptance probability. \n",
    "\n",
    "- [ ] Inputs: \n",
    "    - energy of the current solution (float); \n",
    "    - energy of the neighbour solution (float); \n",
    "    - the current temperature (float); \n",
    "- [ ] Output: probability between 0 and 1 (float).\n",
    "\n",
    "\n",
    "The <small>`math.exp(x)`</small> function can be used to calculate <small>`e`</small> raised to the power <small>`x`</small>.\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "import math \n",
    "\n",
    "def acceptance(c_energy: float, n_energy: float, temp: float) -> float:\n",
    "    ### BEGIN SOLUTION\n",
    "    if n_energy < c_energy:\n",
    "        return 1 \n",
    "    return math.exp(-(n_energy - c_energy)/temp)\n",
    "    ### END SOLUTION\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TESTS \n",
    "# a better solution at the start of our run\n",
    "print(f'expected=1.00, actual={acceptance(c_energy=0.5, n_energy=0.5, temp=1.00):0.2f}')\n",
    "# a worse solution at the start of our run\n",
    "print(f'expected=0.82, actual={acceptance(c_energy=0.5, n_energy=0.7, temp=1.00):0.2f}')\n",
    "# a worse solution in the middle of our run\n",
    "print(f'expected=0.67, actual={acceptance(c_energy=0.5, n_energy=0.7, temp=0.5):0.2f}')\n",
    "# a worse solution at the end of our run\n",
    "print(f'expected=0.00, actual={acceptance(c_energy=0.5, n_energy=0.7, temp=0.001):0.2f}')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Simulated Annealing**\n",
    "\n",
    "Now that we can generate solutions, assess their fitness (energy), and calculate whether to accept a new solution, we can finally perform simulated annealing. \n",
    "\n",
    "<img src=\"https://raw.githubusercontent.com/melbournebioinformatics/COMP90014_2024/master/tutorials/media/week10/pseudocode.png\" width=\"700\"><br>\n",
    "<small>*Pseudocode from the [Wikipedia article](https://en.wikipedia.org/wiki/Simulated_annealing#Pseudocode) on simulated annealing*</small>\n",
    "\n",
    "We will specify a parameter k_max, which controls the number of iterations we will perform. <br> \n",
    "At the end of our run, we will return the current solution (hopefully, the best one). \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div style=\"color: rgb(27,94,32); background: rgb(200,230,201); border: solid 1px rgb(129,199,132); padding: 10px;\">\n",
    "\n",
    "<h3>Exercise 3</h3>\n",
    "    \n",
    "<b>Challenge:</b> Write a function which performs simulated annealing to identify the optimal alignment of two networks. \n",
    "\n",
    "- [ ] Inputs: \n",
    "    - the reference network (nx.Graph); \n",
    "    - the query network (nx.Graph); \n",
    "    - the maximum number of steps k_max (int); \n",
    "- [ ] Output: a tuple containing the final energy and solution\n",
    "\n",
    "Part of the solution has been implemented. \n",
    "You will need to:\n",
    "1. Generate a neighbour solution\n",
    "2. Calculate the energy of this neighbour\n",
    "3. Decide whether to jump using your <small>`acceptance()`</small> function and <small>`random.random()`</small>\n",
    "\n",
    "Note: The <small>`random.random()`</small> function generates a random value between 0 and 1. \n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [],
   "source": [
    "import time \n",
    "import random\n",
    "from typing import Tuple \n",
    "\n",
    "def simulated_annealing(R: nx.Graph, Q: nx.Graph, k_max: int=10000) -> Tuple:\n",
    "    # initialise random starting state and fitness score\n",
    "    c_energy = 1\n",
    "    c_solution = [\n",
    "        ['J', 'K'], ['I', 'A`'], ['H', 'I`'], ['G', 'H`'], ['F', 'G`'],\n",
    "        ['E', 'F`'], ['D', 'E`'], ['C', 'D`'], ['B', 'C`'], ['A', 'B`'], [None, 'J`'],\n",
    "    ]\n",
    "\n",
    "    # iterate\n",
    "    for k in range(k_max):\n",
    "        time.sleep(0.0005)\n",
    "        temp = 1 - (k/k_max)\n",
    "        print(f'temp={temp:0.2f}, energy={c_energy:0.2f}', end='\\r')\n",
    "\n",
    "        # 1. generate neighbour solution\n",
    "        # 2. calculate energy of neighbour solution\n",
    "        # 3. decide whether to jump or stay\n",
    "        ### BEGIN SOLUTION\n",
    "        # generate & assess neighbour solution\n",
    "        n_solution = generate_neighbour(c_solution)\n",
    "        n_energy = objective_function(n_solution, R, Q)\n",
    "\n",
    "        # decide to jump or stay\n",
    "        if acceptance(c_energy, n_energy, temp) >= random.random():\n",
    "            c_energy = n_energy \n",
    "            c_solution = n_solution\n",
    "        ### END SOLUTION\n",
    "\n",
    "    return c_energy, c_solution\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TESTS\n",
    "# feel free to change the time.sleep() value in the function above to increase/decrease run speed\n",
    "energy, solution = simulated_annealing(R, Q)\n",
    "print(f'\\n\\nfinal state:')\n",
    "for r_node, q_node in solution:\n",
    "    print(f'{r_node} -> {q_node}')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "(End of tutorial)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
